lektor/videotools.py

```diff
@@ -0,0 +1,239 @@
+
+    return suffix
+
+
+def get_video_info(filename):
+    """Read video information using ffprobe if available.
+
+    Returns a dict with: width, height and duration.
+    """
+    ffprobe = locate_executable('ffprobe')
+    if ffprobe is None:
+        raise RuntimeError('Failed to locate ffprobe')
+
+    proc = portable_popen([
+        ffprobe,
+        '-v', 'quiet',
+        '-print_format', 'json',
+        '-show_format',
+        '-show_streams',
+        filename,
+    ], stdout=subprocess.PIPE)
+    stdout, _ = proc.communicate()
+
+    if proc.returncode != 0:
+        raise RuntimeError('ffprobe exited with code %d' % proc.returncode)
+
+    ffprobe_data = json.loads(stdout.decode("utf8"))
+    info = {
+        'width': None,
+        'height': None,
+        'duration': None,
+    }
+
+    # Try to extract total video duration
+    try:
+        info['duration'] = timedelta(
+            seconds=float(ffprobe_data['format']['duration']))
+    except (KeyError, TypeError, ValueError):
+        pass
+
+    # Try to extract width and height from the first found video stream
+    for stream in ffprobe_data['streams']:
+        if stream['codec_type'] != 'video':
+            continue
+
+        info['width'] = int(stream['width'])
+        info['height'] = int(stream['height'])
+
+        # We currently don't bother with multiple video streams
+        break
+
+    return info
+
+
+def make_video_thumbnail(ctx, source_video, source_url_path, seek, width,
+                         height=None, crop=False, format=None, quality=None):
+    if format is None:
+        format = 'jpg'
+    if format not in thumbnail_formats:
+        raise ValueError('Invalid thumbnail format')
+
+    # We can't crop if height is undefined
+    if height is None:
+        crop = False
+
+    ffmpeg = locate_executable('ffmpeg')
+    if ffmpeg is None:
+        raise RuntimeError('Failed to locate ffmpeg')
+    info = get_video_info(source_video)
+
+    source_dim = Dimensions(info['width'], info['height'])
+    resize_dim = source_dim.resize(width, height, crop)
+    if crop:
+        target_dim = Dimensions(width, height or resize_dim.height)
+    else:
+        target_dim = resize_dim
+
+    # If there aren't enough pixels we can't resize
+    should_resize = True
+    if source_dim.width <= resize_dim.width or \
+            source_dim.height <= resize_dim.height:
+        should_resize = False
+
+    # Construct a filename suffix unique to the given parameters
+    suffix = get_suffix(seek, width, height, crop=crop, quality=quality)
+    dst_url_path = get_dependent_url(source_url_path, suffix,
+                                     ext='.{}'.format(format))
+
+    if quality is None:
+        quality = 95
+
+    @ctx.sub_artifact(artifact_name=dst_url_path, sources=[source_video])
+    def build_thumbnail_artifact(artifact):
+        artifact.ensure_dir()
+
+        vfilter = 'thumbnail=120,scale={rw}:{rh},crop={tw}:{th}'.format(
```

Comment:
it would be nice if this were customisable, maybe without overriding the ffmpeg default

Reply:
Do you mean the `thumbnail=120` part? I guess I could make that part public as an argument to `frame()`. I would however like to expose as little as possible about the fact that ffmpeg is being used behind the scenes. 120 seemed like a sensible value.

Reply:
Yes. You might consider it low level, but It's useful because it makes a lot of difference on speed and memory consumption vs. relevance. If called "sample" or "sample_size" there wouldn't be anything ffmpeg-specific about it. (its default is 100 btw, close to yours)